src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java

```diff
@@ -0,0 +1,58 @@
+package bio.terra.service.dataset.flight.upgrade.enableSecureMonitoring;
+
+import bio.terra.common.CloudPlatformWrapper;
+import bio.terra.common.exception.FeatureNotImplementedException;
+import bio.terra.common.iam.AuthenticatedUserRequest;
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.dataset.DatasetDao;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.dataset.flight.LockDatasetStep;
+import bio.terra.service.dataset.flight.UnlockDatasetStep;
+import bio.terra.service.job.JobMapKeys;
+import bio.terra.service.policy.PolicyService;
+import bio.terra.service.resourcemanagement.BufferService;
+import bio.terra.service.snapshot.SnapshotService;
+import bio.terra.stairway.Flight;
+import bio.terra.stairway.FlightMap;
+import java.util.UUID;
+import org.springframework.context.ApplicationContext;
+
+public class EnableSecureMonitoringFlight extends Flight {
+
+  public EnableSecureMonitoringFlight(FlightMap inputParameters, Object applicationContext) {
+    super(inputParameters, applicationContext);
+
+    ApplicationContext appContext = (ApplicationContext) applicationContext;
+
+    DatasetService datasetService = appContext.getBean(DatasetService.class);
+    DatasetDao datasetDao = appContext.getBean(DatasetDao.class);
+    SnapshotService snapshotService = appContext.getBean(SnapshotService.class);
+    BufferService bufferService = appContext.getBean(BufferService.class);
+    PolicyService policyService = appContext.getBean(PolicyService.class);
+    FlightMap workingMap = context.getWorkingMap();
+    AuthenticatedUserRequest userReq =
+        inputParameters.get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+
+    UUID datasetId = inputParameters.get(JobMapKeys.DATASET_ID.getKeyName(), UUID.class);
+
+    Dataset dataset = datasetService.retrieve(datasetId);
+    CloudPlatformWrapper platform =
+        CloudPlatformWrapper.of(dataset.getDatasetSummary().getStorageCloudPlatform());
+    if (platform.isGcp()) {
+      addStep(new LockDatasetStep(datasetService, datasetId, false));
+      addStep(
+          new EnableSecureMonitoringFlipFlagStep(
+              datasetId, dataset.isSecureMonitoringEnabled(), datasetDao, userReq));
+      addStep(
+          new EnableSecureMonitoringRefolderGcpProjectsStep(
+              dataset, snapshotService, bufferService, userReq));
+      addStep(
+          new EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep(
+              datasetId, snapshotService, policyService, userReq));
+      addStep(new UnlockDatasetStep(datasetService, datasetId, false));
```

Comment:
Aha!  I just realized, adding a Journal entry recording the switch to secure logging could be a great addition here, even just for the source dataset (probably more fuss than it's worth to journal on each modified snapshot).  That way there will be a record in a dataset's activity log of exactly when secure logging was turned on.

Reply:
There already is one on flipping the flag! [code change here
](https://github.com/DataBiosphere/jade-data-repo/pull/1498/files#diff-a534083fafcf74fa6c64aabdd30c69d5cfc96e296f774302299bf2aa0a4884e8R760)

Do you think that's sufficient?

Reply:
Actually, I think you have the right idea. I'm moving it to its own step at the end of the flight.